contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,248 @@
+
+    // hash of the unbond request => unbond request
+    mapping(bytes32 => UnbondRequest) private _unbondRequests;
+    // user => unbond request queue(hash of the request)
+    mapping(address => DoubleEndedQueueUpgradeable.Bytes32Deque) private _unbondRequestsQueue;
+    // user => locked shares
+    mapping(address => uint256) private _lockedShares;
+    // user => personal unbond sequence
+    mapping(address => uint256) private _unbondSequence;
+
+    struct UnbondRequest {
+        uint256 sharesAmount;
+        uint256 bnbAmount;
+        uint256 unlockTime;
+    }
+
+    /*----------------- events -----------------*/
+    event Delegated(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+    event RewardReceived(uint256 bnbAmount);
+    event UnbondRequested(address indexed sender, uint256 sharesAmount, uint256 bnbAmount, uint256 unlockTime);
+    event UnbondClaimed(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+
+    /*----------------- modifiers -----------------*/
+    modifier whenNotPaused() {
+        require(!IStakeHub(STAKE_HUB_ADDR).isPaused(), "CONTRACT_IS_STOPPED");
+        _;
+    }
+
+    /*----------------- external functions -----------------*/
+    function initialize(address _validator) public payable initializer {
+        validator = _validator;
+
+        assert(msg.value != 0);
+        _bootstrapInitialHolder(msg.value);
+    }
+
+    function delegate(address _delegator) external payable onlyStakeHub whenNotPaused returns (uint256) {
+        return _stake(_delegator, msg.value);
+    }
+
+    function undelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub whenNotPaused returns (uint256) {
+        require(_sharesAmount <= _sharesOf(_delegator), "INSUFFICIENT_BALANCE");
+
+        _lockedShares[_delegator] += _sharesAmount;
+
+        // calculate the BNB amount and update state
+        uint256 _bnbAmount = getPooledBNBByShares(_sharesAmount);
+        _burn(_delegator, _sharesAmount);
+        _totalPooledBNB -= _bnbAmount;
+
+        // add to the queue
+        _unbondSequence[_delegator] += 1; // increase the sequence first to avoid zero sequence
+        bytes32 hash = keccak256(abi.encodePacked(_delegator, _unbondSequence[_delegator]));
+
+        uint256 unlockTime = block.timestamp + IStakeHub(STAKE_HUB_ADDR).getUnbondTime();
+        UnbondRequest memory request = UnbondRequest({sharesAmount: _sharesAmount, bnbAmount: _bnbAmount, unlockTime: unlockTime});
+        _unbondRequests[hash] = request;
+        _unbondRequestsQueue[_delegator].pushBack(hash);
+
+        emit UnbondRequested(_delegator, _sharesAmount, _bnbAmount, request.unlockTime);
+        return _bnbAmount;
+    }
+
+    function redelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub whenNotPaused returns (uint256) {
+        require(_sharesAmount <= _sharesOf(_delegator), "INSUFFICIENT_BALANCE");
```

Comment:
Different from cosmos, we allow A -> B -> C -> D redelegate immediately without limitation, is there any risk?

Reply:
Yes, that's also the logic here. The `redelegate` function in `StakePool` actually just does immediately undelegate action. The whole redelegate process is on `StakeHub`